cwms-data-api/src/main/java/cwms/cda/data/dao/basin/BasinDao.java

```diff
@@ -0,0 +1,179 @@
+        String areaUnitIn = UnitSystem.EN.value().equals(unitSystem)
+                ? Unit.SQUARE_MILES.getValue() : Unit.SQUARE_KILOMETERS.getValue();
+        try {
+            connection(dsl, c -> {
+                setOffice(c, officeId);
+                try (ResultSet rs = basinJooq.catBasins(c, null, null, null, areaUnitIn, officeId)) {
+                    retVal.addAll(buildBasinsFromResultSet(rs, unitSystem));
+                }
+            });
+        } catch (Exception ex) {
+            throw new SQLException(ex);
+        }
+        return retVal;
+    }
+
+    public Basin getBasin(LocationIdentifier basinId, String unitSystem) {
+        CwmsDbBasinJooq basinJooq = new CwmsDbBasinJooq();
+
+        String[] pParentBasinId = new String[1];
+        Double[] pSortOrder = new Double[1];
+        String[] pPrimaryStreamId = new String[1];
+        Double[] pTotalDrainageArea = new Double[1];
+        Double[] pContributingDrainageArea = new Double[1];
+        String pBasinId = basinId.getName();
+        String pOfficeId = basinId.getOfficeId();
+
+        String areaUnitIn = UnitSystem.EN.value().equals(unitSystem)
+                ? Unit.SQUARE_MILES.getValue() : Unit.SQUARE_KILOMETERS.getValue();
+
+        connection(dsl, c -> {
+            setOffice(c, pOfficeId);
+            basinJooq.retrieveBasin(c, pParentBasinId, pSortOrder,
+                    pPrimaryStreamId, pTotalDrainageArea, pContributingDrainageArea, pBasinId,
+                    areaUnitIn, pOfficeId);
+        });
+
+        return new Basin.Builder()
+                .withBasinId(new LocationIdentifier.Builder()
+                        .withName(pBasinId)
+                        .withOfficeId(pOfficeId)
+                        .build())
+                .withTotalDrainageArea(pTotalDrainageArea[0])
+                .withContributingDrainageArea(pContributingDrainageArea[0])
+                .withParentBasinId(new LocationIdentifier.Builder()
+                        .withName(pParentBasinId[0]).
+                        withOfficeId(pOfficeId)
+                        .build())
+                .withSortOrder(pSortOrder[0])
+                .withAreaUnit(areaUnitIn)
+                .withPrimaryStreamId(new LocationIdentifier.Builder()
+                        .withName(pPrimaryStreamId[0])
+                        .withOfficeId(pOfficeId)
+                        .build())
+                .build();
+    }
+
+    public void storeBasin(Basin basin) throws IOException {
+        CwmsDbBasinJooq basinJooq = new CwmsDbBasinJooq();
+
+        basin.validate();
+
+        try {
+            String basinId = basin.getBasinId().getName();
+            String officeId = basin.getBasinId().getOfficeId();
+            String parentBasinId = basin.getParentBasinId() != null ? basin.getParentBasinId().getName() : null;
+            Double sortOrder = basin.getSortOrder();
+            String primaryStreamId = basin.getPrimaryStreamId().getName();
```

Comment:
I believe primaryStreamID can be null. Null check this (and anywhere else something similar is done)

Reply:
Addressed with a null check for now until Streams are fully implemented